pkg/koordlet/runtimehooks/protocol/pod_context.go

```diff
@@ -122,15 +124,45 @@ func (p *PodRequest) FromReconciler(podMeta *statesinformer.PodMeta) {
 	}
 }
 
+type RecorderEvent struct {
+	HookName  string
+	MsgFmt    string
+	Reason    string
+	EventType string
+}
+
 type PodResponse struct {
 	Resources Resources
 }
 
 type PodContext struct {
-	Request  PodRequest
-	Response PodResponse
-	executor resourceexecutor.ResourceUpdateExecutor
-	updaters []resourceexecutor.ResourceUpdater
+	Request        PodRequest
+	Response       PodResponse
+	executor       resourceexecutor.ResourceUpdateExecutor
+	updaters       []resourceexecutor.ResourceUpdater
+	RecorderEvents []RecorderEvent
+}
+
+func (p *PodContext) RecordEvent(r record.EventRecorder, pod *corev1.Pod) {
+	// Noraml, Warning => RecordEvent
+	events := make(map[string]RecorderEvent)
+	for _, event := range p.RecorderEvents {
+		if !recutil.ValidateEventType(event.EventType) {
+			klog.Warningf("EventType is not valid %v", event)
+			continue
+		}
+
+		e := event
+		if _, ok := events[event.EventType]; ok {
+			e.MsgFmt += "-" + event.MsgFmt
+			e.Reason += "-" + event.Reason
+		}
+		events[event.EventType] = e
```

Comment:
how about send event here to keep the order

Reply:
do you mean to issue several events here instead of merge them ?

Reply:
Sorry, I missed the thing that the events are grouped by types.

Reply:
yes, already done with it.